File: worker/src/pipeline/generate-markdown.ts
import type { ProcessedSection } from "../types.js";

interface MarkdownInput {
  videoId: string;
  title: string;
  channelName: string;
  duration: number;
  language: string;
  sections: ProcessedSection[];
}

/**
 * Generates a markdown file string from processed sections and metadata.
 * Output follows the spec in docs/transcript-format.md:
 * YAML frontmatter + body with <!-- t:SECONDS --> comments before ## headings.
 * Called after LLM processing, before uploading to storage.
 */
export function generateMarkdown(input: MarkdownInput): string {
  const { videoId, title, channelName, duration, language, sections } = input;

  const frontmatterSections = sections.map((s) => ({
    title: s.title,
    timestamp: s.timestamp,
  }));

  const frontmatter = [
    "---",
    `video_id: "${videoId}"`,
    `title: "${escapeYaml(title)}"`,
    `channel: "${escapeYaml(channelName)}"`,
    `duration: ${duration}`,
    `language: "${language}"`,
    `sections:`,
    ...frontmatterSections.map(
      (s) => `  - title: "${escapeYaml(s.title)}"\n    timestamp: ${s.timestamp}`,
    ),
    "---",
  ].join("\n");

  const body = sections
    .map((s) => `<!-- t:${s.timestamp} -->\n## ${s.title}\n\n${s.content}`)
    .join("\n\n");

  return `${frontmatter}\n\n${body}\n`;
}

function escapeYaml(str: string): string {
  return str.replace(/"/g, '\\"');
}

File: worker/src/pipeline/generate_markdown.py
from __future__ import annotations

from src.models import ProcessedSection


def generate_markdown(
    *,
    video_id: str,
    title: str,
    channel_name: str,
    duration: int,
    language: str,
    sections: list[ProcessedSection],
) -> str:
    """
    Build a Markdown file from processed sections and metadata.
    Output follows docs/transcript-format.md: YAML frontmatter + body with
    <!-- t:SECONDS --> comments before ## headings.
    """
    fm_sections = "\n".join(
        f'  - title: "{_escape_yaml(s.title)}"\n    timestamp: {s.timestamp}'
        for s in sections
    )

    frontmatter = (
        "---\n"
        f'video_id: "{video_id}"\n'
        f'title: "{_escape_yaml(title)}"\n'
        f'channel: "{_escape_yaml(channel_name)}"\n'
        f"duration: {duration}\n"
        f'language: "{language}"\n'
        f"sections:\n"
        f"{fm_sections}\n"
        "---"
    )

    body = "\n\n".join(
        f"<!-- t:{s.timestamp} -->\n## {s.title}\n\n{s.content}"
        for s in sections
    )

    return f"{frontmatter}\n\n{body}\n"


def _escape_yaml(text: str) -> str:
    """Escape double quotes inside YAML string values."""
    return text.replace('"', '\\"')
